File: analytics/src/main/java/io/appmetrica/analytics/impl/component/ComponentMigrationHelper.java
package io.appmetrica.analytics.impl.component;

import androidx.annotation.VisibleForTesting;
import java.util.LinkedList;
import java.util.List;

public class ComponentMigrationHelper {

    public static class Creator {

        ComponentMigrationHelper create() {
            return new ComponentMigrationHelper();
        }
    }

    private List<MigrationScript> mMigrationScripts;

    private ComponentMigrationHelper() {
        initMigrationScripts();
    }

    private void initMigrationScripts() {
        mMigrationScripts = new LinkedList<>();
    }

    void checkMigration() {
        for (MigrationScript script : mMigrationScripts) {
            script.checkMigration();
        }
    }

    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
    List<MigrationScript> getMigrationScripts() {
        return mMigrationScripts;
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/component/ComponentMigrationHelper.kt
package io.appmetrica.analytics.impl.component

import androidx.annotation.VisibleForTesting
import io.appmetrica.analytics.logger.appmetrica.internal.DebugLogger

internal class ComponentMigrationHelper private constructor(componentUnit: ComponentUnit) {

    private val tag = "[ComponentMigrationHelper]"

    class Creator {
        fun create(componentUnit: ComponentUnit): ComponentMigrationHelper {
            return ComponentMigrationHelper(componentUnit)
        }
    }

    private val migrationScripts: List<ComponentMigrationScript> = listOf(
        ComponentMigrationToV113(componentUnit),
    )

    fun migrate(from: Int) {
        this.migrationScripts.forEach {
            DebugLogger.info(tag, "run migration: `${it.javaClass}` from apiLevel = `$from`")
            it.run(from)
        }
    }

    @VisibleForTesting
    fun getMigrationScripts(): List<ComponentMigrationScript> = migrationScripts
}
